### PR TITLE
Explicitly check for duplicate entry status when inserting entry

### PR DESCRIPTION
**Pull request by nupuruttarwar**
Signed-off-by: Nupur Uttarwar <nupur.uttarwar@intel.com>

### CODE REVIEW

File: stratum/hal/lib/tdi/tdi_sde_table_entry.cc [hunk 1]
```diff
@@ -65,10 +65,12 @@ ::util::Status TdiSdeWrapper::InsertTableEntry(
...
 
   const auto flags = ::tdi::Flags(0);
-  RETURN_IF_TDI_ERROR(table->entryAdd(
+  tdi_status_t status = table->entryAdd(
       *real_session->tdi_session_, *dev_tgt, flags, *real_table_key->table_key_,
-      *real_table_data->table_data_))
-      << "Could not add table entry with: " << dump_args();
+      *real_table_data->table_data_);
+  if (!((status == BF_SUCCESS) || (status == BF_ALREADY_EXISTS)))
+     return ::util::Status(::util::error::Code::INTERNAL,
```
Comment (ffoulkes) [inline stratum/hal/lib/tdi/tdi_sde_table_entry.cc]: Why do this instead of  `return MAKE_ERROR(code) << message`?

Comment (nupuruttarwar) [reply stratum/hal/lib/tdi/tdi_sde_table_entry.cc]: Right, done

File: stratum/hal/lib/tdi/tdi_sde_table_entry.cc [hunk 2]
```diff
@@ -65,11 +65,13 @@ ::util::Status TdiSdeWrapper::InsertTableEntry(
    */
 
   const auto flags = ::tdi::Flags(0);
-  RETURN_IF_TDI_ERROR(table->entryAdd(
+  tdi_status_t status = table->entryAdd(
```
Comment (ffoulkes) [inline stratum/hal/lib/tdi/tdi_sde_table_entry.cc:68]: Are you sure that this is the correct behavior in all cases? What if the caller attempts to add an entry with the same key but different data?

On the other hand, if the caller is expected to have verified the non-existence of the entry before issuing the Add, then failure could reasonably be considered an internal error.

If the answer is "it depends", then this method should probably return ::util::error::Code::ALREADY_EXISTS and allow the caller to make the decision (which might require edits at one or more of the call sites).

(Apologies if I've already asked this. This change looks vaguely familiar.)

File: stratum/hal/lib/tdi/tdi_sde_table_entry.cc [hunk 3]
```diff
@@ -139,9 +141,12 @@ ::util::Status TdiSdeWrapper::DeleteTableEntry(
   device->createTarget(&dev_tgt);
 
   const auto flags = ::tdi::Flags(0);
-  RETURN_IF_TDI_ERROR(table->entryDel(
-      *real_session->tdi_session_, *dev_tgt, flags, *real_table_key->table_key_))
-      << "Could not delete table entry with: " << dump_args();
+  tdi_status_t status = table->entryDel(
```
Comment (ffoulkes) [inline stratum/hal/lib/tdi/tdi_sde_table_entry.cc:144]: Are you sure that this is the correct behavior in all cases?

It is not necessarily wrong to treat a key-not-found condition as a success when you're deleting a key. The caller is asserting that there should be no table entry with the specified key, and that condition is true.

On the other hand, if the caller is expected to have verified the existence of the key before issuing the Delete, failure would reasonably be considered an internal error.

If the answer is "it depends", then perhaps this method should return ::util::error::Code::NOT_FOUND and allow the caller to make the decision (which might require edits at one or more of the call sites).

(Apologies if I've already asked this.)

File: stratum/hal/lib/tdi/tdi_sde_table_entry.cc (hunk 2 above)
Comment (nupuruttarwar) [reply stratum/hal/lib/tdi/tdi_sde_table_entry.cc:68]: Yes, this is correct behavior. If, table key is same and data/action associated with it has changed, then user is expected to use ::p4::v1::Update::Type as MODIFY and  ModifyTableEntry SDE wrapper API will be invoked. I am not sure, if there is support for this in the backend. I will check with SDE team and if yes, then we can print the message accordingly.

**Review by ffoulkes — APPROVED** [review]
I still have my doubts, but I will defer to your judgment here.

**Pull request merged**